### PR TITLE
Fixed air unit resupply

### LINKED ISSUE #5966: Implement queued reloading for aircraft [$69]
Since aircraft need to be reloaded, it is very time consuming and boring to reload 20 helicopters at one helipad for example. But if there was a feature that reloaded each aircraft one after another when a group of them are ordered to reload, it would make this process much easier. The same logic could also be applied to repairs at the service depot.

<bountysource-plugin>
---
There is a **[$69 open bounty](https://www.bountysource.com/issues/3148314-implement-queued-reloading-for-aircraft?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bateramos**
Correction to the resupply queue for airplanes and helicopters.

https://www.youtube.com/watch?v=VSiwaujBaPA

Fixes  #5966.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs [hunk 1]
```diff
@@ -36,12 +36,12 @@ public FlyAttack(Actor self, Target target)
 		public override Activity Tick(Actor self)
 		{
 			if (!target.IsValidFor(self))
-				return NextActivity;
+				return new ReturnToBase(self);
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Why?

File: OpenRA.Mods.Common/Activities/Execute.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#endregion
+
+using System;
+using OpenRA.Activities;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	public class Execute : Activity
+	{
+		Action f;
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Activities/Execute.cs]: We used to have an activity like this, called 'CallFunction'. It was nuked from orbit.

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs [hunk 3]
```diff
@@ -20,6 +20,8 @@ namespace OpenRA.Mods.Common.Activities
 {
 	public class ReturnToBase : Activity
 	{
+		static int ticksForNextValidation = 50;
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs]: This should come from traitinfo data.

File: OpenRA.Mods.Common/Activities/Execute.cs (hunk 2 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Execute.cs]: In this case, I should create a Activity that wait for the airport to receive a reservation call?

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs (hunk 1 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: This is to make the aircraft try to return to the resupply base even if its already reserved after destroying a target. It should be better to preserve the NextActivity? Maybe but it on a Activity Sequence that the ReturnToBase is the last?
I know for sure that when there is no more commands, NextActivity is null. Maybe only return to base is this case?

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs (hunk 3 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs]: ok!

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: These will break queued activity chains (where players can queue orders using the shift key).

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 4]
```diff
@@ -44,16 +44,33 @@ public override Activity Tick(Actor self)
...
 				else
-					return ActivityUtils.SequenceActivities(new HeliFly(self, Target.FromActor(nearestHpad)));
+				{
+					var distanceFromHelipad = (nearestHpad.CenterPosition - self.CenterPosition).HorizontalLength;
+					var distanceLength = heli.Info.WaitDistanceFromResupplyBase.Length;
+
+					// If no pad is available, move near one and wait
+					if (distanceFromHelipad > distanceLength)
+					{
+						var randomXPosition = self.World.SharedRandom.Next(-distanceLength, distanceLength);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: You can use `WVec.FromPDF` to get a random 2D offset (then multiply by `distanceLength / 1024`) instead of DIYing your own.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 5]
```diff
@@ -44,16 +44,33 @@ public override Activity Tick(Actor self)
...
+
+						return ActivityUtils.SequenceActivities(new HeliFly(self, target, WDist.Zero, heli.Info.WaitDistanceFromResupplyBase), this);
+					}
+
+					return this;
+				}
 			}
 
 			heli.MakeReservation(dest);
+			heli.DisableRepulsing();
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: This looks dangerous.  What do we gain by manually disabling and enabling repulsation?

File: OpenRA.Mods.Common/Activities/Execute.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Execute.cs]: This would be best handled by the long-discussed and long-procrastinated docking rewrite.  CC @reaperrr and @penev92.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 6]
```diff
@@ -176,6 +189,18 @@ public void Repulse()
 			SetPosition(self, CenterPosition + FlyStep(speed, repulsionForce.Yaw.Facing));
 		}
 
+		public void DisableRepulsing()
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: This looks dangerous.  I would strongly prefer it if we didn't do this – we have wasted so much time over the years debugging issues with manually managed state like this.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs (hunk 1 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: But if I call SequenceActivities with ReturnToBase as the last activity will fix this problem?

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs (hunk 4 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Ok, I will change that!

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: It should work if you use `QueueActivity` instead.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 6 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: The other alternative is creating a manager that control the queue and the positions of the helicopters to resupply. I could not find any other solution for this. The repulsion from others helicopters sometime knocks the helicopter that was descending away from the supply pad.
I see as two options. Create this controller that handle all positioning of the helicopters or disable the repulsion (maybe there should be a better way of disabling this).
But i really all hear to what do you thing is better. In the case of the controller, think I need a example class that I can base my work or just a north on how this should be done.

File: OpenRA.Mods.Common/Activities/Execute.cs (hunk 2 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Execute.cs]: Maybe I can help on this too. There is a forum thread or a issue on this?

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 6 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: There is already a check in GetRepulsionForce that disables the repulsion if the actor isn't at its cruising altitude.  I suspect the problem is that it gets repulsed during the one tick where it is at cruising height and preparing to land.  You could change that check to additionally disable repulsion if the actor is within X WRange of the reserved docking point.

Comment (bateramos) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: I made some changes in the last commit. I think your approach solved this issue.

File: OpenRA.Mods.Common/Activities/Execute.cs (hunk 2 above)
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Execute.cs]: Actually, the `CallFunc` activity still exists.

Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Execute.cs]: But I can use this Class? Its used in 13 different places.

Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Execute.cs]: Yes, you can.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs [hunk 7]
```diff
@@ -32,8 +32,17 @@ public override Activity Tick(Actor self)
...
 
+			if (aircraft.IsPlane)
+				return ActivityUtils.SequenceActivities(
+					aircraft.GetResupplyActivities(host)
+					.Append(new CallFunc(() => aircraft.UnReserve()))
+					.Append(new WaitFor(() => Reservable.IsReserved(host)))
+					.Append(new TakeOff(self))
+					.Append(NextActivity).ToArray());
+
+			// If is helicopter move alway as soon as the resupply ends
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: Typo:

```patch
-alway
+away
```

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: OpenRA.Mods.Common/Activities/Air/FlyCircleTimed.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2014 The OpenRA Developers (see AUTHORS)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Activities/Air/FlyCircleTimed.cs]: 2016 and GPLv3 or later. https://github.com/OpenRA/OpenRA/pull/10801

Comment (Mailaender) [inline OpenRA.Mods.Common/Activities/Air/FlyCircleTimed.cs]: You may also want to add yourself to `AUTHORS`.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs [hunk 9]
```diff
@@ -36,12 +36,15 @@ public FlyAttack(Actor self, Target target)
...
+			{
+				self.QueueActivity(new ReturnToBase(self));
 				return NextActivity;
+			}
 
 			// Move to the next activity only if all ammo pools are depleted and none reload automatically
 			// TODO: This should check whether there is ammo left that is actually suitable for the target
 			if (ammoPools.All(x => !x.Info.SelfReloads && !x.HasAmmo()))
-				return NextActivity;
+				return new ReturnToBase(self);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: This is going to erase the order queue.  You need to sequence the RTB before continuing the NextActivity with something like `return ActivityUtils.SequenceActivities(new ReturnToBase(self), NextActivity)`.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs [hunk 10]
```diff
@@ -36,12 +36,15 @@ public FlyAttack(Actor self, Target target)
 		public override Activity Tick(Actor self)
 		{
 			if (!target.IsValidFor(self))
+			{
+				self.QueueActivity(new ReturnToBase(self));
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: This still doesn't look right, unfortunately  This appends the RTB at the end of the activity queue, which in theory should mean it is redundant vs the `ReturnOnIdle` trait.  If another activity is blocking then neither is going to run.

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs [hunk 11]
```diff
@@ -32,8 +32,17 @@ public override Activity Tick(Actor self)
 			if (host == null)
 				return NextActivity;
 
+			if (aircraft.IsPlane)
+				return ActivityUtils.SequenceActivities(
+					aircraft.GetResupplyActivities(host)
+					.Append(new CallFunc(() => aircraft.UnReserve()))
+					.Append(new WaitFor(() => Reservable.IsReserved(host)))
+					.Append(new TakeOff(self))
+					.Append(NextActivity).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: SequenceActivities accepts a list as parameters, so building your own array shouldn't be needed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 12]
```diff
@@ -83,6 +83,12 @@ public class AircraftInfo : ITraitInfo, IPositionableInfo, IFacingInfo, IOccupyS
 		[Desc("Sound to play when the actor is landing.")]
 		public readonly string LandingSound = null;
 
+		[Desc("The distance of the resupply base that the aircraft will wait for its turn.")]
+		public readonly WDist WaitDistanceFromResupplyBase = new WDist(3072);
+
+		[Desc("The number of ticks that a airplane will wait to make a new search for a available airport.")]
+		public readonly int NumberOfTicksToVerifyAvailableAirport = 50;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: This is only two seconds at normal game speed.  Isn't that a bit fast?

**Comment by reaperrr** [comment]
Also needs a rebase.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 12 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: `an available`

File: OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs (hunk 11 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs]: I used this way to reutilize the aircraft GetSupplyActivities, that return a IEnumerable<Activity>

**Comment by bateramos** [comment]
sorry @reaperrr why i need to rebase? Is the commit message?

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 13]
```diff
@@ -44,13 +44,28 @@ public override Activity Tick(Actor self)
...
 				if (nearestHpad == null)
 					return ActivityUtils.SequenceActivities(new Turn(self, initialFacing), new HeliLand(self, true), NextActivity);
 				else
-					return ActivityUtils.SequenceActivities(new HeliFly(self, Target.FromActor(nearestHpad)));
+				{
+					var distanceFromHelipad = (nearestHpad.CenterPosition - self.CenterPosition).HorizontalLength;
+					var distanceLength = heli.Info.WaitDistanceFromResupplyBase.Length;
+
+					// If no pad is available, move near one and wait
+					if (distanceFromHelipad > distanceLength)
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Shouldn't helicopters also move *away* from helipads when they're too close to make room for others?

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 14]
```diff
@@ -181,6 +188,13 @@ public virtual WVec GetRepulsionForce()
 			if (!Info.Repulsable)
 				return WVec.Zero;
 
+			if (reservation != null)
+			{
+				var distanceFromReservationActor = (reservedActor.CenterPosition - self.CenterPosition).HorizontalLength;
+				if (distanceFromReservationActor < Info.WaitDistanceFromResupplyBase.Length)
+					return WVec.Zero;
+			}
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: I'm not sure if I like this (mind you, I haven't playtested it yet, so it might be alright). My first gut instinct would be to separate waiting helicopters vertically instead of disabling repulsion. Like a holding pattern, kinda.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 15]
```diff
@@ -44,13 +44,28 @@ public override Activity Tick(Actor self)
 			{
 				var rearmBuildings = heli.Info.RearmBuildings;
 				var nearestHpad = self.World.ActorsHavingTrait<Reservable>()
-									.Where(a => a.Owner == self.Owner && rearmBuildings.Contains(a.Info.Name))
-									.ClosestTo(self);
+					.Where(a => a.Owner == self.Owner && rearmBuildings.Contains(a.Info.Name))
+					.ClosestTo(self);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Eventually we might to add the behaviour the harvesters use for refineries now (ie, three harvs per refinery, if it's full, move on to the next one).

**Comment by obrakmann** [comment]
In `HeliReturnToBase` (after line 80), you need to add `NextActivity` behind `ResupplyAircraft`.

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs [hunk 16]
```diff
@@ -108,7 +108,10 @@ public override Activity Tick(Actor self)
 
 				self.CancelActivity();
 				if (nearestAfld != null)
-					return ActivityUtils.SequenceActivities(new Fly(self, Target.FromActor(nearestAfld)), new FlyCircle(self));
+					return ActivityUtils.SequenceActivities(
+						new Fly(self, Target.FromActor(nearestAfld), WDist.Zero, plane.Info.WaitDistanceFromResupplyBase),
+						new FlyCircleTimed(plane.Info.NumberOfTicksToVerifyAvailableAirport, self),
+						new ReturnToBase(self));
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs]: Use `this` instead of creating a new instance. I had better results with that change. The activity queue probably gets lost otherwise.

**Comment by obrakmann** [comment]
This looks very nice indeed. This just needs a couple of changes to make sure the activity queue doesn't get lost, then it's good to go. :+1: 

I'll still need to check how this impacts scripted missions, though.

Rebasing this won't be very hard, git manages to auto-merge the changes, so you don't actually have to do anything besides running the rebase command.

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs [hunk 17]
```diff
@@ -108,7 +108,10 @@ public override Activity Tick(Actor self)
 
 				self.CancelActivity();
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs]: This isn't actually needed.

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs [hunk 18]
```diff
@@ -108,7 +108,10 @@ public override Activity Tick(Actor self)
...
 				if (nearestAfld != null)
-					return ActivityUtils.SequenceActivities(new Fly(self, Target.FromActor(nearestAfld)), new FlyCircle(self));
+					return ActivityUtils.SequenceActivities(
+						new Fly(self, Target.FromActor(nearestAfld), WDist.Zero, plane.Info.WaitDistanceFromResupplyBase),
+						new FlyCircleTimed(plane.Info.NumberOfTicksToVerifyAvailableAirport, self),
+						new ReturnToBase(self));
 				else
 					return new FlyCircle(self);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs]: Please change this to `return NextActivity`. Otherwise (some of) the mission scripts break, and 2) if `NextActivity` is null, aircraft will end up circling by themselves anyway, thanks to `FlyAndContinueWithCirclesWhenIdle`.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs (hunk 13 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Hmmmm I don't think this is necessary. In my tests, the problem occurs when the helicopters were approaching to resupply. When are fully supplied, they already form a line in the helipad exit waypoint.

Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Hmmmm I don't think this is necessary. In my tests, the problem occurs when the helicopters were approaching to resupply. When are fully supplied, they already form a line in the helipad exit waypoint.

Comment (obrakmann) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: I was thinking of helicopters that are already there for other reasons, because the player ordered them there for example, not because they were sent there for rearming. But I agree that it's a corner case that might not actually matter in the end.

Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: I just made a test with 10 helicopters on top of the helipad, indeed the heli that just resupplied get "stuck" with the others, but this does not prevent others heli from resupply, as on approach the repulsion is disabled. Its not "elegant" but it is funcional.

**Comment by bateramos** [comment]
Sorry to disturbe again with this merge messages, but how I can remove this from my branch?

**Comment by obrakmann** [comment]
Just run a rebase once, it will remove the merge commit.

Looks good to me now, :+1:

**Comment by obrakmann** [comment]
This unfortunately needs a rebase again. I'd hold off from doing that until this gets the second :+1:, though. I won't add the tag either to keep people from intentionally skipping this pr.

**Comment by Mailaender** [comment]
Still works as promised. :white_check_mark:

**Comment by bateramos** [comment]
the rebase is now ok?
Checksum error.. What! Why? How?

**Comment by Mailaender** [comment]
Our Windows CI https://ci.appveyor.com/project/OpenRA/openra/history is unstable for some reason. Don't worry. The build went fine. Just the setup.exe creation failed, because https://chocolatey.org/packages/nsis didn't install properly.

**Comment by Mailaender** [comment]
I hope https://github.com/OpenRA/OpenRA/pull/11644 will help in the future.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bdc6e4397091d77a8b7e7989313d0a4f5c521eab)
